Issue 145: Fix aggregate-add-third-party modifying project dependencies

srdo opened the pull request:
Fixes https://github.com/mojohaus/license-maven-plugin/issues/145.

The issue occurs because the plugin modifies the MavenProject dependency lists in DefaultDependencyTool.loadProjectArtifacts. Dependencies may in some cases be removed if they belong to the same reactor as the current project.  

I've added a modified version of the reproduction code from https://github.com/jimonthebarn/maven-license-plugin-issue as a test.

The changes are attempting to preserve the current behavior of the plugin, while also not modifying the MavenProject fields.

ppalaga commented on src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java:
```diff
@@ -559,6 +561,12 @@
      */
     private Map<String, String> globalKnownLicenses;
+    
+    boolean isAggregatorBuild;
```
Can `isAggregatorBuild` ever be false? Both subclasses set it to `true` or I oversee something? Can perhaps `isAggregatorBuild` be eliminated through encapsulation?

srdo replied:
It is false when using `AddThirdPartyPlugin` as a goal. It is only true when the plugin is called via `AggregateAddThirdPartyMojo`. The line in `AddThirdPartyPlugin` where it is set to true is only called through `AggregateAddThirdPartyMojo` (it essentially invokes the plugin once for each child project in the reactor).

Regarding whether it can be eliminated, I don't think so without a larger refactoring. The problem is that the `init` method needs to behave differently when used as part of an `AggregateAddThirdPartyMojo` invocation, both when called directly on the `AggregateAddThirdPartyMojo` class, and the `AddThirdPartyMojo` instances it creates. `init` doesn't take parameters, and the interface is defined at the `AbstractLicenseMojo` level. I didn't want to start rewriting these classes to not use inheritance so heavily, so adding a field instead was the only solution I could think of.

ppalaga replied:
> It is false when using AddThirdPartyPlugin as a goal.

I see `isAggregatorBuild = true` in `AddThirdPartyPlugin` https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-f71923bddd2327bb3c7b3c62c80878c2R467

> Regarding whether it can be eliminated, I don't think so without a larger refactoring. The problem is that the init method needs to behave differently when used as part of an AggregateAddThirdPartyMojo invocation, both when called directly on the AggregateAddThirdPartyMojo class, and the AddThirdPartyMojo instances it creates. init doesn't take parameters, and the interface is defined at the AbstractLicenseMojo level.

Not sure I understand all details, but can't you define something like `protected abstract LoadedProjectDependencies loadDependencyArtifacts()` in `AbstractAddThirdPartyMojo`, implement it differently in the two subclasses and replace this conditional https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-0bfb2b7e70b9abacb4339555c7b49e81R668 with the follwing? 

```
dependencyArtifacts = loadDependencyArtifacts()
```

ppalaga replied:
BTW, if you really intended to have isAggregatorBuild = false in AddThirdPartyPlugin https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-f71923bddd2327bb3c7b3c62c80878c2R467 and all tests are passing, then there is probably a test missing that would fail due to this mistake?

srdo replied:
> I see isAggregatorBuild = true in AddThirdPartyPlugin https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-f71923bddd2327bb3c7b3c62c80878c2R467

Yes. This line is in the `initFromMojo` method, which is used by `AggregatorAddThirdPartyMojo` to initialize the `AddThirdPartyMojo` instances it creates (here: https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-38b03aa7a6c189ea0f7e814313a14682R215). It isn't used when running `AddThirdPartyMojo` as a goal directly.

This is also why doing something like `protected abstract LoadedProjectDependencies loadDependencyArtifacts()` doesn't work. The behavior of `loadDependencyArtifacts()` would need to do one thing when `AddThirdPartyMojo` runs it as part of the `addThirdParty` goal, while it needs to do something else when `AggregatorAddThirdPartyMojo` instantiates a bunch of `AddThirdPartyMojo`s to run for its child projects.

Trying to give an overview:
* Running the `addThirdParty` goal from Maven: `init` has to act as if `isAggregatorBuild == false`. 
* Running `aggregateAddThirdParty` will run `init` once through `AggregatorAddThirdPartyMojo` for the parent project, where `isAggregatorBuild == true`, and then run `init` once per child project through `AddThirdPartyMojo.initFromMojo` where `isAggregatorBuild == true`.

ppalaga replied:
> Yes. This line is in the `initFromMojo` method, which is used by `AggregatorAddThirdPartyMojo` to initialize the `AddThirdPartyMojo` instances it creates (here: https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-38b03aa7a6c189ea0f7e814313a14682R215). It isn't used when running `AddThirdPartyMojo` as a goal directly.

OK, thanks for the explanation, I was not aware of the fact that `AggregatorAddThirdPartyMojo`  instantiates `AddThirdPartyMojo`s.

Given what you say, `isAggregatorBuild` seems to be significant for `AddThirdPartyMojo` only (because it is always true for `AggregatorAddThirdPartyMojo`). Could we perhaps move `isAggregatorBuild` to `AddThirdPartyMojo` and introduce a `protected abstract LoadedProjectDependencies loadDependencyArtifacts()` in `AbstractAddThirdPartyMojo` and let  it alwas return

* `dependenciesTool.loadProjectArtifacts( localRepository, remoteRepositories, project, reactorProjectDependencies )` from `AggregatorAddThirdPartyMojo` and
* the current conditional from `AddThirdPartyMojo`? 

I think this would improve the readability of the code.

srdo replied:
Made this change. Turns out the method wasn't necessary in the aggregator mojo, so I removed it from there.

ppalaga commented on src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java:
```diff
@@ -559,6 +561,12 @@
+    
+    boolean isAggregatorBuild;
+    
+    Map<String, List<Dependency>> reactorProjectDependencies;
```
Please explain in JavaDoc what is this field good for.

srdo replied:
Will do

ppalaga commented on src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java:
```diff
@@ -559,6 +561,12 @@
+    
+    Map<String, List<Dependency>> reactorProjectDependencies;
+    
+    LoadedProjectDependencies dependencyArtifacts;
```
Please explain in JavaDoc what is this field good for.

ppalaga commented on src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java:
```diff
@@ -634,6 +642,12 @@ protected void init()
             doGenerateBundle = false;
         }
+        
+        if (isAggregatorBuild) {
```
Please apply the code style of the project (if you choose to keep this).

ppalaga commented on src/main/java/org/codehaus/mojo/license/AbstractDownloadLicensesMojo.java:
```diff
@@ -263,7 +264,9 @@ protected MavenProject getProject()
     {
-        return dependenciesTool.loadProjectDependencies( project, this, localRepository, remoteRepositories, null );
+        return dependenciesTool.loadProjectDependencies( 
+                new LoadedProjectDependencies(project.getArtifacts(), project.getDependencyArtifacts()),
```
Code style

ppalaga commented on src/main/java/org/codehaus/mojo/license/AggregatorAddThirdPartyMojo.java:
```diff
@@ -157,6 +157,11 @@ protected void init() throws Exception {
+        isAggregatorBuild = true;
+        reactorProjectDependencies = new TreeMap<>();
+        for (MavenProject reactorProject : this.reactorProjects) {
+            reactorProjectDependencies.put(String.format("%s:%s", reactorProject.getGroupId(), reactorProject.getArtifactId()), reactorProject.getDependencies());
```
Code style

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/LoadedProjectDependencies.java:
```diff
@@ -0,0 +1,28 @@
+
```
Missing file header

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/LoadedProjectDependencies.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Set;
+import org.apache.maven.artifact.Artifact;
+
+public class LoadedProjectDependencies {
```
Missing JavaDoc

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/LoadedProjectDependencies.java:
```diff
@@ -0,0 +1,28 @@
+    private final Set<Artifact> allDependencies;
+    private final Set<Artifact> directDependencies;
+
+    public LoadedProjectDependencies( Set<Artifact> allDependencies, Set<Artifact> directDependencies )
```
I vote for making this class immutable. Either copy the constructor params and wrap them in an Collections.unmodifiableSet or find another way to reach the same effect.

srdo replied:
Will do.

srdo commented on src/main/java/org/codehaus/mojo/license/api/LoadedProjectDependencies.java:
```diff
@@ -0,0 +1,63 @@
+
+    public Set<Artifact> getAllDependencies()
+    {
+        return Collections.unmodifiableSet( allDependencies );
```
Ok, will change

srdo commented:
I think there's a small bug in `DefaultDependenciesTool.loadProjectDependencies`, will try to make a test that covers it and update.

srdo commented:
Rebased, and fixed another few bugs in `loadProjectArtifacts`. I wound up replacing the method. The previous implementation would fail to find transitive dependencies for long chains of sibling dependencies in the same reactor, and would use the wrong list of remote repositories for looking up dependencies. Both issues are demonstrated in it/ISSUE-145-2. The latter bug would cause the build to crash. 

The issue with wrong transitive dependencies wasn't a big deal, but I think we might as well fix it. When the aggregator mojo finds dependencies for sibling projects, it was possible to be missing some dependencies in the return from `loadProjectArtifacts`. For instance if you have siblings A, B, C and A depends on B depends on C, then `loadProjectArtifacts` for A would replace B with C in the direct dependency list. This was wrong, but since the aggregator mojo processes all modules in the reactor, B would still be picked up when the aggregator mojo moved on to processing B.

srdo commented:
The Java 7 tests failed, they pass for me locally. Going to retrigger the build.

ppalaga commented:
> Rebased, and fixed another few bugs in `loadProjectArtifacts`. I wound up replacing the method. The previous implementation would fail to find transitive dependencies for long chains of sibling dependencies in the same reactor, and would use the wrong list of remote repositories for looking up dependencies. Both issues are demonstrated in it/ISSUE-145-2. 

Thanks a lot for the fixes and special thanks for the test!

> The issue with wrong transitive dependencies wasn't a big deal, but I think we might as well fix it.

Not sure you imply (a) you fixed it or (b) it needs to be fixed. If (b) is the case, please file a new issue.

The PR looks good to me now. Please squash the commits, and reword to something like `Fix #145 aggregate-add-third-party modifying project dependencies`  and I'll merge afterwards.

srdo commented:
> Not sure you imply (a) you fixed it or (b) it needs to be fixed. If (b) is the case, please file a new issue.

It is fixed now. 

Squashed.

ppalaga commented:
@srdo good work, thanks!
